### DISCUSSION TITLE
Combox : open dropdown on click

### DISCUSSION THREAD

**Discussion by ootsun** [tailwindlabs/headlessui#1205] [Help]
Hello,
I'm using combobox and I'd like that a click in the input causes the dropdown to open.
Is this comportement already implemented and I'm doing something wrong or is it normal that nothing happens on click?
![image](https://user-images.githubusercontent.com/25452154/157121622-b4f3dbd2-2365-49ab-8ecf-14cfda6df2ab.png)

**Comment by ootsun** [comment]
I found a workaround : 

`const [openOptions, setOpenOptions] = useState(false);`

then use onFocus and onBlur on a div inside the Combobox

```
<Combobox value={selectedPerson} onChange={setSelectedPerson}>
                  <div  onFocus={() => setOpenOptions(true)}  onBlur={() => setOpenOptions(false)}>
```

and finally

`<Combobox.Options  static={openOptions}`

**Reply by anubhav-sl** [reply]
@ootsun Great find.

How did you handle selection of an option. When I select an option onBlur is triggered first, and because of that I am not able to get an option selected.

**Reply by ootsun** [reply]
Yeah, you are right. Try using "onClick" instead of "onFocus" and let me know if it's solve your problem please. Will update workaround if you confirm "onClick" is the right approche.

**Reply by eakl** [reply]
onClick doesn't work either

**Reply by jonbellah** [reply]
Ran into this same issue. I was able to workaround the `onBlur` / item selection race condition by throwing a timeout on the close... not the most elegant solution but seems to be working okay so far.

`onBlur={() => setTimeout(() => setOpen(false), 100)}`

**Reply by wenis** [reply]
I know this is an old thread, but for search engine purposes, I'd like to provide a solution for selecting any default text in the input and opening the dropdown on click for vue3.

Create a ref for the ComboButton:

`const comboboxButtonRef = ref()`

In your template:

`<ComboboxButton ref="comboboxButtonRef" class="..." />`

Use v-on:click in the ComboboxInput component to handle the event method...

`<ComboboxInput @click="onComboButtonClick($event, open)" ... />`

And in your script handle the event...

```
  function onComboButtonClick(e, open) {
    e.target.select()
    if (!open) comboboxButtonRef.value.el.click()
  }
```

IMO, selecting placeholder text and opening combobox list items is the most user friendly way of handling this type of input.

**Reply by arprinceofficial** [reply]
> I know this is an old thread, but for search engine purposes, I'd like to provide a solution for selecting any default text in the input and opening the dropdown on click for vue3.
> 
> Create a ref for the ComboButton:
> 
> `const comboboxButtonRef = ref()`
> 
> In your template:
> 
> `<ComboboxButton ref="comboboxButtonRef" class="..." />`
> 
> Use v-on:click in the ComboboxInput component to handle the event method...
> 
> `<ComboboxInput @click="onComboButtonClick($event, open)" ... />`
> 
> And in your script handle the event...
> 
> ```
>   function onComboButtonClick(e, open) {
>     e.target.select()
>     if (!open) comboboxButtonRef.value.el.click()
>   }
> ```
> 
> IMO, selecting placeholder text and opening combobox list items is the most user friendly way of handling this type of input.

thank you... I have been searching for long time... Best wishes @wenis

**Comment by anubhav-sl** [comment]
Multiple timeouts gets added in this case the options closes unexpectedly.

**Comment by dimalakh** [comment]
my workaround for this issue:
1. make ref of `Combobox.Button`
2. emulate Button click while clicking on Input
`onClick={() => buttonRef.current?.click()}`

**Reply by ayhanap** [reply]
I used your workaround with added a check on render prop.

`onClick={() => { if (!open) openCloseButtonRef.current?.click(); }}`

**Reply by cosileone** [reply]
For those of us using typescript here's how you make a ref that's `.click`able:
`const comboboxButtonRef = useRef<HTMLButtonElement>(null);`

**Comment by knitevision1** [comment]
I approached it this way:

1. Wrap `<Combobox.Button><Combobox.Input /><Combobox.Button>`
2. Button will handle opening on click, while Combobox input would remain in control for `onFocus`, `onBlur` events

**Reply by richardstrnad** [reply]
Amazing, thank you!

```
<ComboboxButton>
  <ComboboxInput
    @change="query = $event.target.value"
    :displayValue="(c) => c.name"
  />
</ComboboxButton>
```

**Reply by ayhanap** [reply]
Easy solution but clicking on input multiple times closes the combobox, which is a bit strange UX.

I used solution with ref on [this comment](https://github.com/tailwindlabs/headlessui/discussions/1205#discussioncomment-3074325) but added a condition with render prop `open`

`onClick={() => {
                      if (!open) openCloseButtonRef.current?.click();
                    }}`

**Reply by jeffjenk** [reply]
I found another option for avoiding the closing issue when clicked again, just stop the click propagation on the input:
```
<Combobox.Button>
  {({ open }) => (
    <Combobox.Input
      onChange={event => setQuery(event.target.value)}
      onClick={e => {
        if (open) e.stopPropagation();
      }}
    />
  )}
</Combobox.Button>
```

**Reply by ThienLE0811** [reply]
Amazing, thank you!

**Reply by navinrangar** [reply]
thanks mate!

**Reply by Odysseus14** [reply]
For this solution, does inputting a ‘space’, does it close the control?

**Reply by tusmenkoe** [reply]
This leads to a opening dropdown during scroll, as touch events triggers the button click.

**Comment by dbousamra** [comment]
Hmm. What about if I want to open the Combobox when clicking the input, but only if it's not already open

**Reply by tembra** [reply]
> Hmm. What about if I want to open the Combobox when clicking the input, but only if it's not already open

You must check if it's already opened with `open` render prop from [`<Combobox>` component](https://headlessui.com/react/combobox#combobox).

I used the same approach as @dimalakh [posted here](https://github.com/tailwindlabs/headlessui/discussions/1205#discussioncomment-3074325), only adding this verification I just mention.

**Comment by AnthonyAtSigma2** [comment]
I think a reasonable feature would be to allow people to control the state of the open property without forcing them to re-create the internal state changes that come out of the box. For example, I want to be able to open the ComboBox programmatically, but I don't want to lose all the other neat features that already cause the ComboBox to automatically open or close.

It seems like the closest thing that we have is the ability to open the ComboBox by ".click"-ing it, and there are some really good threads above discussing how to do that.

However, this is obviously a pretty brittle workaround, since it depends on including the `ComboBox.Button` element. It also doesn't give you an easy approach for closing the thing programmatically. I'm sure someone will chime in with a way to close it by clicking on an element outside of the `ComboBox`, but I think this is still very brittle.

The easiest solution I can think of from a design perspective is exposing some render props called `openFn` and `closeFn`, which can be used to go through Headless UI's state change doors, so to speak. Then the `open` prop continues to be a `readonly` property which can be used for the most common use cases, but also can be used for this more complicated version to reflect the internal state.

**Comment by alessiolilli** [comment]
Headlessui already merged a PR, but they haven't been released yet. If you install the @insiders version , you can use a props named 'immediate' that should do exactly what you need

**Reply by flkrnr** [reply]
btw this is the mentioned PR: https://github.com/tailwindlabs/headlessui/pull/2686

**Comment by dheerajsinghnagdali** [comment]
In my opinion, the best way to achieve this is rendering `div` in the `as` prop of the button and using the input inside it. Moreover, you could use a simple `span` for rendering the icon.

**Comment by nemo369** [comment]
I updated my package to version 2.0.3 and now the `useRef` and click() hack doesn't work. Does Someone have another workaround maybe?

**Reply by alessiolilli** [reply]
they added the immediate props inside combox so we don't need to use workaround anymore

**Comment by dylanbromby** [comment]
Is the immediate prop React only?

Edit: yes.

**Reply by dylanbromby** [reply]
Well, my hack for this in the Vue component is to use the nullable prob and send a space character into the input when clicked which opens the menu, then immediately clear the space character, leaving the menu open. Not the prettiest, but the use of the space character is imperceptible to me; it happens very, very quickly.
